### PR TITLE
implemented RegistrationService and added unit tests by means of TDD

### PR DESCRIPTION
**Pull request by ivan0dyatlyukkk**
…principle

### CODE REVIEW

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(emptyCharacterLoginUser);
+        });
+
+        User sixWhiteSpacesLoginUser = new User(1L, INVALID_LOGIN_EXAMPLE5,
+                VALID_PASSWORD_EXAMPLE,
+                MIN_VALID_AGE);
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(sixWhiteSpacesLoginUser);
+        });
```
Comment (holsoni) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:130]: I think that you don't have to check every possible length of login. 

>  Few not-ok checks you can have in one not-ok test method. Same for ok-test. Testing only ‘abc’ (3-length - not-ok) and ‘abcdefgh’ (8-length - ok) is not enough because we do not check how the program behave with edge cases.

You can check if it is less than 6 symblos, if it equals and is more than 6 and if it is null.

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+        User actual = null;
+        try {
+            actual = registrationService.register(expected);
+        } catch (RegistrationException e) {
+            fail(e.getMessage());
+        }
+        assertEquals(expected, actual);
+        int expectedSizeOfStorage = INITIAL_SIZE_OF_STORAGE + 1;
+        int actualSizeOfStorage = Storage.people.size();
+        assertEquals(expectedSizeOfStorage, actualSizeOfStorage);
```
Comment (holsoni) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:58]: When you test creation of some objects, it is better to check if the object is in the  storage using method .get(). 
```suggestion
    void register_newUser_ok() {
        User expected = new User(1L, VALID_LOGIN_EXAMPLE, VALID_PASSWORD_EXAMPLE, MIN_VALID_AGE);
        registrationService.register(expected);
        int expectedSizeOfStorage = INITIAL_SIZE_OF_STORAGE + 1;
        int actualSizeOfStorage = Storage.people.size();
        assertEquals(expectedSizeOfStorage, actualSizeOfStorage);
        User actual = storageDaoImpl.get(expected.getUsernam();
        assertEquals(expected, actual);
        
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+        User sixWhiteSpacesPasswordUser = new User(1L, VALID_LOGIN_EXAMPLE,
+                INVALID_PASSWORD_EXAMPLE5,
+                MIN_VALID_AGE);
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(sixWhiteSpacesPasswordUser);
+        });
+
+        int actualSizeOfStorage = Storage.people.size();
+        assertEquals(INITIAL_SIZE_OF_STORAGE, actualSizeOfStorage);
+    }
```
Comment (holsoni) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:200]: Same as with login length test.

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+        try {
+            actual = registrationService.register(expected);
+        } catch (RegistrationException e) {
+            fail(e.getMessage());
+        }
+        assertEquals(expected, actual);
+        int expectedSizeOfStorage = INITIAL_SIZE_OF_STORAGE + 1;
+        int actualSizeOfStorage = Storage.people.size();
+        assertEquals(expectedSizeOfStorage, actualSizeOfStorage);
+    }
```
Comment (holsoni) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:225]: Same as with the register_validUser test

**Review by holsoni — CHANGES_REQUESTED** [review]

**Review by holsoni — APPROVED** [review]

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        });
+        int actualSizeOfStorage = Storage.people.size();
+        assertEquals(INITIAL_SIZE_OF_STORAGE, actualSizeOfStorage);
+    }
+
+    @Test
+    void register_userWithValidLogin_ok() {
+        User expected = new User(VALID_LOGIN_EXAMPLE, VALID_PASSWORD_EXAMPLE, MIN_VALID_AGE);
+        registrationService.register(expected);
+        User actual = storageDao.get(expected.getLogin());
```
Comment (andrii-hoienko) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:75]: It is better to get by login

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+                VALID_PASSWORD_EXAMPLE,
+                MIN_VALID_AGE);
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(fourCharacterLoginUser);
+        });
+
+        User oneCharacterLoginUser = new User(INVALID_LOGIN_EXAMPLE3,
+                VALID_PASSWORD_EXAMPLE,
+                MIN_VALID_AGE);
+        assertThrows(RegistrationException.class, () -> {
```
Comment (andrii-hoienko) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:98]: It is better to check error message as well

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+    @Test
+    void register_userWithInvalidLogin_notOk() {
+        User fiveCharacterLoginUser = new User(INVALID_LOGIN_EXAMPLE1,
+                VALID_PASSWORD_EXAMPLE,
+                MIN_VALID_AGE);
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(fiveCharacterLoginUser);
+        });
+
+        User fourCharacterLoginUser = new User(INVALID_LOGIN_EXAMPLE2,
```
Comment (andrii-hoienko) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:88]: do you really need 4 and 5 character cases?

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+import core.basesyntax.db.Storage;
+import core.basesyntax.exceptions.RegistrationException;
+import core.basesyntax.model.User;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class RegistrationServiceImplTest {
+    private static final User INITIAL_USER_IN_STORAGE = new User("validLogin",
```
Comment (andrii-hoienko) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:17]: Is this really needed?

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+    @Test
+    void register_userWithValidPassword_ok() {
+        User expected = new User(VALID_LOGIN_EXAMPLE, VALID_PASSWORD_EXAMPLE, MIN_VALID_AGE);
+        registrationService.register(expected);
+        User actual = storageDao.get(expected.getLogin());
+        assertEquals(expected, actual);
+    }
+
+    @Test
+    void register_userWithInvalidPassword_notOk() {
```
Comment (andrii-hoienko) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:125]: to much asserts for test, to check invalid password one wrong user will be enough(check other your test as well)

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    @Test
+    void register_userWithLessThanMinAge_notOk() {
+        User user = new User(VALID_LOGIN_EXAMPLE, VALID_PASSWORD_EXAMPLE, INVALID_AGE_EXAMPLE);
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(user);
+        });
+        int actualSizeOfStorage = Storage.people.size();
+        assertEquals(INITIAL_SIZE_OF_STORAGE, actualSizeOfStorage);
```
Comment (andrii-hoienko) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:176]: could be extracted to method and reused

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Good, but let`s improve your solution

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java (hunk 10 above)
Comment (ivan0dyatlyukkk) [reply src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:176]: Frankly, I have just rethought and figured out that these were redundant because we didn't test the working of the Storage.
In this case, I figured it out correctly, didn't I?

File: src/main/java/core/basesyntax/service/RegistrationService.java [hunk 11]
```diff
@@ -1,7 +1,8 @@
 package core.basesyntax.service;
 
+import core.basesyntax.exceptions.RegistrationException;
 import core.basesyntax.model.User;
 
 public interface RegistrationService {
-    User register(User user);
+    User register(User user) throws RegistrationException;
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/service/RegistrationService.java:7]: should we indicate that method throws RuntimeException?

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 12]
```diff
@@ -2,13 +2,66 @@
...
 
 public class RegistrationServiceImpl implements RegistrationService {
+    private static final int MIN_LOGIN_PASSWORD_NUMBER_OF_CHARACTERS = 6;
+    private static final int MIN_AGE = 18;
     private final StorageDao storageDao = new StorageDaoImpl();
 
     @Override
-    public User register(User user) {
-        return null;
+    public User register(User user) throws RegistrationException {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:15]: same

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+import core.basesyntax.db.Storage;
+import core.basesyntax.exceptions.RegistrationException;
+import core.basesyntax.model.User;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class RegistrationServiceImplTest {
+    private static User INITIAL_USER_IN_STORAGE;
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:17]: why name written in upper case? this is not a constant

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+        assertEquals(NULL_LOGIN_ERROR_MESSAGE, thrown.getMessage());
+    }
+
+    @Test
+    void register_userWithValidPassword_ok() {
+        User expected = new User(VALID_LOGIN_EXAMPLE, VALID_PASSWORD_EXAMPLE, MIN_VALID_AGE);
+        registrationService.register(expected);
+        User actual = storageDao.get(VALID_LOGIN_EXAMPLE);
+        assertEquals(expected, actual);
+    }
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:127]: same

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+        assertEquals(NULL_PASSWORD_ERROR_MESSAGE, thrown.getMessage());
+    }
+
+    @Test
+    void register_userWithValidAge_ok() {
+        User expected = new User(VALID_LOGIN_EXAMPLE, VALID_PASSWORD_EXAMPLE, MIN_VALID_AGE);
+        registrationService.register(expected);
+        User actual = storageDao.get(VALID_LOGIN_EXAMPLE);
+        assertEquals(expected, actual);
+    }
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:166]: same

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java (hunk 13 above)
Comment (ivan0dyatlyukkk) [reply src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:17]: Ohh, sorry, my bad, I forgot to rename it after the last change

File: src/main/java/core/basesyntax/service/RegistrationService.java (hunk 11 above)
Comment (ivan0dyatlyukkk) [reply src/main/java/core/basesyntax/service/RegistrationService.java:7]: Yeap, you're right. In my first version, I made a checked exception and after changing I didn't fix it

**Review by KseniiaMakarova — APPROVED** [review]
